### PR TITLE
Javalab - Add light mode theme and theme toggle

### PR DESCRIPTION
**Pull request by JillianK**
Added a light mode theme and the ability to toggle between light and dark modes. Final designs are still TBD

https://user-images.githubusercontent.com/17147070/113945523-52471200-97bb-11eb-8c97-3b1306e5e626.mov

## Links

- jira ticket: [CSA-203](https://codedotorg.atlassian.net/browse/CSA-203?atlOrigin=eyJpIjoiODI1YWQ4ZjI3YjMxNDhkZmFjZTBhYjZkYjgyZGI1Y2UiLCJwIjoiaiJ9)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/javalab/JavalabView.jsx [hunk 1]
```diff
@@ -75,7 +75,8 @@ class JavalabView extends React.Component {
 
   state = {
     loading: true,
-    loadSuccess: null
+    loadSuccess: null,
+    isDarkMode: false
```
Comment (molly-moen) [inline apps/src/javalab/JavalabView.jsx]: would it be worth putting `isDarkMode` in redux so we don't have to pass it down to other components? We'll be adding more components as we add more features that will need to know the theme.

File: apps/src/javalab/JavalabView.jsx [hunk 2]
```diff
@@ -151,18 +183,20 @@ class JavalabView extends React.Component {
                 </button>
               </div>
               <div style={style.buttons}>
-                <button
-                  type="button"
-                  style={style.singleButton}
-                  onClick={this.compile}
+                <JavalabSettings
+                  style={{
```
Comment (molly-moen) [inline apps/src/javalab/JavalabView.jsx]: can we extract the button styling out into a variable/function so we don't repeat the code for each button?

**Review by molly-moen — COMMENTED** [review]
this looks great!

File: apps/src/javalab/JavalabEditor.jsx [hunk 3]
```diff
@@ -35,6 +35,8 @@ const style = {
   }
 };
 
+const darkTabColor = '#282c34';
```
Comment (jmkulwik) [inline apps/src/javalab/JavalabEditor.jsx]: Suggestion: Move this to a `darkBackground` or `darkMode` style i.e. 
```
darkBackground: {
  backgroundColor: '#282c34`
}
```
Then merge the styles like you do in JavalabConsole.jsx. This would allow us to easily add additional styling in the future and be sure it will be applied everywhere that is relevant.

**Review by molly-moen — APPROVED** [review]
LGTM!

File: apps/src/javalab/JavalabEditor.jsx [hunk 4]
```diff
@@ -61,16 +64,38 @@ class JavalabEditor extends React.Component {
...
       }),
       parent: this._codeMirror,
       dispatch: this.dispatchEditorChange()
     });
   }
 
+  componentDidUpdate(prevProps) {
+    if (prevProps.isDarkMode !== this.props.isDarkMode) {
+      if (this.props.isDarkMode) {
+        this.editor.dispatch({
```
Comment (jmkulwik) [inline apps/src/javalab/JavalabEditor.jsx:90]: Should this be using `dispatchEditorChange`?

File: apps/src/javalab/JavalabSettings.jsx [hunk 5]
```diff
@@ -0,0 +1,118 @@
+import React, {Component} from 'react';
```
Comment (jmkulwik) [inline apps/src/javalab/JavalabSettings.jsx:1]: Nice component! The toggle logic is very easy to follow and the error handling is clear and consistent. This will make the component easy to maintain.

File: apps/src/javalab/JavalabView.jsx [hunk 6]
```diff
@@ -114,7 +116,35 @@ class JavalabView extends React.Component {
...
+    return [
+      <a onClick={this.props.toggleDarkMode}>
+        Switch to {isDarkMode ? 'light mode' : 'dark mode'}
+      </a>
+    ];
+  };
+
+  getButtonStyles = isSettings => {
+    const {isDarkMode} = this.props;
+    if (isDarkMode) {
```
Comment (jmkulwik) [inline apps/src/javalab/JavalabView.jsx:130]: Suggestion: Change this to `if ... else if ... else` - That will prevent future maintainers from accidentally inserting statements between the three conditions.

File: apps/src/javalab/JavalabView.jsx [hunk 7]
```diff
@@ -114,7 +116,35 @@ class JavalabView extends React.Component {
...
+    if (isSettings) {
+      return {...style.singleButton, backgroundColor: color.orange};
+    }
+    return {...style.singleButton, backgroundColor: color.cyan};
+  };
+
   renderJavalab() {
+    const {isDarkMode} = this.props;
+    if (isDarkMode) {
+      document.getElementsByTagName('body')[0].style.backgroundColor =
```
Comment (jmkulwik) [inline apps/src/javalab/JavalabView.jsx]: NIT: Replace `document.getElementsByTagName('body')[0]` with `document.body`

File: apps/src/javalab/JavalabView.jsx [hunk 8]
```diff
@@ -151,18 +186,12 @@ class JavalabView extends React.Component {
                 </button>
               </div>
               <div style={style.buttons}>
+                <JavalabSettings style={this.getButtonStyles(true)}>
```
Comment (jmkulwik) [inline apps/src/javalab/JavalabView.jsx]: NIT: Since `true` is a primitive, it's difficult to understand what it represents. I've seen people mitigate this in the past by adding a comment i.e. `this.getButtonStyles(true /* isSettingsButton */)`

File: apps/src/javalab/JavalabView.jsx [hunk 9]
```diff
@@ -114,7 +116,35 @@ class JavalabView extends React.Component {
...
+  renderSettings = () => {
+    const {isDarkMode} = this.props;
+    return [
+      <a onClick={this.props.toggleDarkMode}>
+        Switch to {isDarkMode ? 'light mode' : 'dark mode'}
+      </a>
+    ];
+  };
+
+  getButtonStyles = isSettings => {
```
Comment (jmkulwik) [inline apps/src/javalab/JavalabView.jsx]: Super minor NIT: I didn't understand what `isSettings` was until I read how `getButtonStyles` was used. Perhaps `isSettingsButton` would be clearer?

File: apps/src/javalab/JavalabSettings.jsx [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+        }
+      });
+    }
+  };
+
+  state = {
+    dropdownOpen: false
+  };
+
+  expandDropdown = () => {
```
Comment (jmkulwik) [inline apps/src/javalab/JavalabSettings.jsx:58]: This component would benefit from some tests coverage of the various state changes to future proof as we add more items to settings.

File: apps/src/javalab/JavalabEditor.jsx [hunk 11]
```diff
@@ -61,16 +64,38 @@ class JavalabEditor extends React.Component {
...
-        extensions: editorSetup
+        extensions: extensions
       }),
       parent: this._codeMirror,
       dispatch: this.dispatchEditorChange()
     });
   }
 
+  componentDidUpdate(prevProps) {
+    if (prevProps.isDarkMode !== this.props.isDarkMode) {
```
Comment (jmkulwik) [inline apps/src/javalab/JavalabEditor.jsx:88]: This is the "mainline" scenario for changing between dark and light mode. Adding a unit test here will ensure this change persists well into the future.

File: apps/src/javalab/JavalabConsole.jsx [hunk 12]
```diff
@@ -130,7 +135,13 @@ class JavalabConsole extends React.Component {
         <PaneHeader hasFocus={true}>
           <PaneSection>Console</PaneSection>
         </PaneHeader>
-        <div style={style.consoleStyle} ref={el => (this._consoleLogs = el)}>
+        <div
+          style={{
+            ...style.consoleStyle,
+            ...(this.props.isDarkMode ? style.darkMode : style.lightMode)
```
Comment (jmkulwik) [inline apps/src/javalab/JavalabConsole.jsx:141]: Given the tight visual coupling between the console and editor, losing the ability to change between dark and light in the console in the future would be confusing to the user. Adding a unit test here would help prevent that.

File: apps/src/javalab/JavalabView.jsx (hunk 9 above)
Comment (jmkulwik) [inline apps/src/javalab/JavalabView.jsx]: NIT: It would be nice to have some test coverage here as well to future proof the dark/light mode transition.

File: apps/src/javalab/JavalabEditor.jsx (hunk 4 above)
Comment (JillianK) [reply apps/src/javalab/JavalabEditor.jsx:90]: I think that's for real editor changes while this is for extension changes according to the Configuration section on [this page](https://codemirror.net/6/docs/migration/)

Comment (jmkulwik) [reply apps/src/javalab/JavalabEditor.jsx:90]: 👍

File: apps/test/unit/javalab/JavalabSettingsTest.js [hunk 13]
```diff
@@ -0,0 +1,50 @@
+import {assert} from '../../util/deprecatedChai';
```
Comment (jmkulwik) [inline apps/test/unit/javalab/JavalabSettingsTest.js]: should be `reconfiguredChai`

File: apps/test/unit/javalab/JavalabSettingsTest.js [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
+  it('is initially just a button', () => {
+    const wrapper = shallow(<JavalabSettings {...defaultProps} />);
+    assert.strictEqual(wrapper.children().length, 1);
+    assert.strictEqual(wrapper.childAt(0).name(), 'button');
+  });
+
+  it('shows children when clicked', () => {
+    const wrapper = shallow(<JavalabSettings {...defaultProps} />);
+    wrapper.find('button').simulate('click');
```
Comment (jmkulwik) [inline apps/test/unit/javalab/JavalabSettingsTest.js]: In general, avoid `simulate('click')` on shallow wrapped enzyme elements. It's [discouraged by the writers of enzyme](https://github.com/enzymejs/enzyme/issues/2173). They recommend calling the event handler directly. i.e.
`wrapper.find('button').instance().toggleDropdown()` you may need to call `wrapper.update()` afterward.

File: apps/test/unit/javalab/JavalabViewTest.js [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+import pageConstants, {setPageConstants} from '@cdo/apps/redux/pageConstants';
+import {
+  getStore,
+  registerReducers,
+  stubRedux,
+  restoreRedux
+} from '@cdo/apps/redux';
+import javalab, {toggleDarkMode} from '@cdo/apps/javalab/javalabRedux';
+import color from '@cdo/apps/util/color';
+global.$ = require('jquery');
```
Comment (jmkulwik) [inline apps/test/unit/javalab/JavalabViewTest.js:9]: Curious: Why is this needed? I would have expected jquery to already exist here.

Comment (JillianK) [reply apps/test/unit/javalab/JavalabViewTest.js:9]: JavalabView gave me an error when I tried to run the tests without it so I guess it didn't exist

**Review by jmkulwik — APPROVED** [review]
Great work!

File: apps/test/unit/javalab/JavalabViewTest.js [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+    );
+  };
+
+  describe('getButtonStyles', () => {
+    it('Is cyan or orange in light mode', () => {
+      const editor = createWrapper();
+      const notSettings = editor
+        .find('JavalabView')
+        .instance()
+        .getButtonStyles(false);
```
Comment (JillianK) [inline apps/test/unit/javalab/JavalabViewTest.js:59]: This line is making MakerApp tests fail?

File: apps/test/unit/javalab/JavalabViewTest.js [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+      onCommitCode: () => {},
+      isProjectLevel: false,
+      isReadOnlyWorkspace: false,
+      isDarkMode: false
+    };
+  });
+
+  describe('getButtonStyles', () => {
+    it('Is cyan or orange in light mode', () => {
+      let editor = shallow(<JavalabView {...defaultProps} />);
```
Comment (jmkulwik) [inline apps/test/unit/javalab/JavalabViewTest.js:29]: Also leave a comment here or on line 3 explaining the issues with mount - especially that it loads many shared DOM components which cause collisions with other tests. This will allow future code maintainers to avoid the same long investigation you had to do. It's not immediately obvious that JavalabView loads such a large section of the page, or that it would cause an issue with something as seemingly unrelated as the Circuit Playground board.

File: apps/test/unit/javalab/JavalabViewTest.js [hunk 18]
```diff
@@ -0,0 +1,41 @@
+import React from 'react';
+import {expect} from '../../util/reconfiguredChai';
+import {shallow} from 'enzyme';
+// We use the UnconnectedJavalabView because shallow().dive() does not work with our version of react-redux
```
Comment (jmkulwik) [inline apps/test/unit/javalab/JavalabViewTest.js]: Keep it to what we know. i.e. we know that calling shallow().dive().instance() on the connected JavalabView fails to give access to the methods owned by JavalabView. You could also say we think it has something to do with the very old version of react-redux we're using. If you do, specify the version so a future dev can check the behavior when the version gets updated. Without a bug report from Enzyme, React, or Redux, however, we don't know this for sure. For example, it's also possible that our manual mocking of the redux store (via apps/src/redux.js) is causing these issues. On the other hand, if you're able to find a bug report describing this behavior, link it here.

A future code maintainer will likely pick up our investigation at some point. Comments like these are intended to give as much information as possible without leading them in the wrong direction.

File: apps/src/javalab/JavalabView.jsx [hunk 19]
```diff
@@ -192,13 +221,17 @@ class JavalabView extends React.Component {
   }
 }
 
+// We use the UnconnectedJavalabView to make this component testable with our version of react-redux.
```
Comment (jmkulwik) [inline apps/src/javalab/JavalabView.jsx]: Also add in a note that this is a deprecated pattern. It's pretty common to copy/paste a component as a starting point for a new component. Given the fragility this pattern lends to tests, it'd be good to warn people away from this unless it meets their specific needs.

Optional: Make the details around why we're using it here more specific. This would allow someone who sees this in the future reason about whether using this pattern would be appropriate for their use case.

**Comment by jmkulwik** [comment]
Already signed off, but here's my ✔️ for the new set of changes!
